### ISSUE TITLE
Optional type is not narrowed in for loop with continue

### ISSUE THREAD

**Issue by Andarius** [microsoft/pyright#2801]
**Describe the bug**
`Optional[T]` type of a dict value is not narrowed to T when checking `is None` inside a for loop with `continue`

It raises the following error.
```python
test.py:17:15 - error: Argument of type "list[int] | None" cannot be assigned to parameter "__obj" of type "Sized" in function "len"
    Type "list[int] | None" cannot be assigned to type "Sized"
      "__len__" is not present (reportGeneralTypeIssues)
```

**To Reproduce**
Run `pyright` on the code below

**Expected behavior**
Should not raise an error

**Screenshots or Code**

```python
from typing import Optional

my_dict: dict[str, Optional[list[int]]] = {
    'foo': [1, 2],
    'bar': None
}

for key in my_dict.keys():
    value = my_dict[key]
    if value is None:
        continue
    print(len(value))  # No Error

for key in my_dict.keys():
    if my_dict[key] is None:
        continue
    print(len(my_dict[key]))  # Raise Error
```

**Comment by erictraut** [comment]
Index expressions with dynamic subscript expressions are never narrowed. In general, narrowing is unsafe here because the container can be mutated in ways that are not obvious to a type checker. For example, the container could be extended (e.g. elements inserted), the element corresponding to the subscript expression could be modified, etc.

None of the type checkers that I'm familiar with (including those in other languages like TypeScript) support type narrowing of index expressions with dynamic subscripts.

One safe workaround is to do the following:

```python
for key in my_dict.keys():
    if (x := my_dict[key]) is None:
        continue
    print(len(x))
```
